python/dgl/contrib/sampling/dis_sampler.py

```diff
@@ -0,0 +1,114 @@
+# This file contains DGL distributed samplers APIs.
+from ...network import _send_subgraph, _recv_subgraph
+from ...network import _batch_send_subgraph, _batch_recv_subgraph
+from ...network import _create_sampler_sender, _create_sampler_receiver
+from ...network import _finalize_sampler_sender, _finalize_sampler_receiver
+
+class SamplerSender(object):
+    """Sender of DGL distributed sampler.
+
+    Users use SamplerSender class to send sampled 
+    subgraph (NodeFlow) to remote trainer. Note that, SamplerSender
+    class will try to connect to SamplerReceiver in a loop until the
+    SamplerReceiver started.
+
+    Parameters
+    ----------
+    ip : str
+        ip address of remote trainer machine
+    port : int
+        port of remote trainer machine
+    """
+    def __init__(self, ip, port):
+        self._ip = ip
+        self._port = port
+        self._sender = _create_sampler_sender(ip, port)
+
+    def __del__(self):
+        """Finalize Sender
+        """
+        # _finalize_sampler_sender will send a special message
+        # to tell the remote trainer machine that it has finished its job.
+        _finalize_sampler_sender(self._sender)
+
+    def Send(self, nodeflow):
+        """Send sampled subgraph (NodeFlow) to remote trainer.
+
+        Parameters
+        ----------
+        nodeflow : NodeFlow
+            sampled NodeFlow object
+        """
+        _send_subgraph(self._sender, nodeflow)
+
+    def BatchSend(self, nodeflow_list):
+        """Send a batch of sampled subgraph (NodeFlow) to remote trainer.
+
+        Parameters
+        ----------
+        nodeflow_list : list
+            A list of NodeFlow objects
+        """
+        _batch_send_subgraph(self._sender, nodeflow_list)
+
+class SamplerReceiver(object):
+    """Receiver of DGL distributed sampler.
+
+    Users use SamplerReceiver class to receive sampled 
+    subgraph (NodeFlow) from remote samplers. Note that SamplerReceiver 
+    can receive messages from multiple senders concurrently, by given 
+    the num_sender parameter, and only when all senders connect to SamplerReceiver,
+    the SamplerReceiver can start its job.
+
+    Parameters
+    ----------
+    ip : str
+        ip address of current trainer machine
+    port : int
+        port of current trainer machine
+    num_sender : int
+        total number of sampler nodes, use 1 by default
+    """
+    def __init__(self, ip, port, num_sender=1):
```

Comment:
do users have to provide `num_sender` if more than one sender is used?

Reply:
Yes